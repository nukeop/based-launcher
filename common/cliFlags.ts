export type CLIFlagKeys = "theme" | "inputPrefix";

export type CLIFlags = {
  [key in CLIFlagKeys]: string;
};
