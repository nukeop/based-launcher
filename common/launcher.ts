export enum LauncherActionType {
  Execute = "execute",
  RunDesktopFile = "run-desktop-file",
  Return = "return",
}

// This tells the renderer what to do with the selected item
export type LauncherAction = {
  type: LauncherActionType;
  payload?: string;
};

export type LauncherOption = {
  id: string;
  name: string;
  description?: string;
  icon?: string | null;
  onAction: LauncherAction;
};
