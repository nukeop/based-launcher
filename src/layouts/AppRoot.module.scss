div.app-root {
    position: relative;
    width: 100%;
    height: 100%;
    border-radius: 0;

    background: rgba(0, 0, 0, 0.3);
    border-radius: 24px;
}