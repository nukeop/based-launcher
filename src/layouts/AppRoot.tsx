import styles from "./AppRoot.module.scss";
import { Paper } from "@mui/material";

type AppRootProps = {
  children: React.ReactNode;
};

export const AppRoot: React.FC<AppRootProps> = ({ children }) => (
  <Paper elevation={0} className={styles["app-root"]}>
    {children}
  </Paper>
);
