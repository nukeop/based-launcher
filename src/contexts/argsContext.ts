import { CLIFlags } from "../../common/cliFlags";
import { DesktopEntry } from "../../common/desktop-entries";
import { createContext } from "react";

type Args = {
  stdinArgs?: string[];
  cliFlags?: CLIFlags;
  desktopEntries?: DesktopEntry[];
};

export const ArgsContext = createContext<Args>({
  stdinArgs: [],
  cliFlags: {},
  desktopEntries: [],
});
