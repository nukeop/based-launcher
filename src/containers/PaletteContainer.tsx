import { IpcEvent } from "../../common/ipc";
import { Palette } from "../components/Palette/Palette";
import { ArgsContext } from "../contexts/argsContext";
import { ipcRenderer } from "electron";
import Fuse from "fuse.js";
import { useCallback, useContext, useEffect, useState } from "react";

export const PaletteContainer: React.FC = () => {
  const { stdinArgs } = useContext(ArgsContext);
  const options =
    stdinArgs?.map((arg, index) => ({
      id: arg,
      name: arg,
      icon: null,
      onAction: () => onAction(arg),
    })) ?? [];

  const onAction = useCallback((item: string) => {
    ipcRenderer.send(IpcEvent.ReturnSelectedItem, item);
  }, []);

  const [filterInput, setFilterInput] = useState("");
  const [selectedItemIndex, setSelectedItemIndex] = useState(0);

  const fuse = new Fuse(options, {
    keys: ["name", "description"],
    threshold: 0.3,
  });

  const filteredOptions = filterInput
    ? fuse.search(filterInput).map((result) => result.item)
    : options;

  const onUp = useCallback(
    (event: KeyboardEvent) => {
      if (event.key === "ArrowUp") {
        setSelectedItemIndex(
          selectedItemIndex === 0
            ? filteredOptions.length - 1
            : selectedItemIndex - 1
        );
      }
    },
    [filteredOptions, selectedItemIndex]
  );

  const onDown = useCallback(
    (event: KeyboardEvent) => {
      if (event.key === "ArrowDown") {
        setSelectedItemIndex(
          selectedItemIndex === filteredOptions.length - 1
            ? 0
            : selectedItemIndex + 1
        );
      }
    },
    [filteredOptions, selectedItemIndex]
  );

  const onEnter = useCallback(
    (event: KeyboardEvent) => {
      if (event.key === "Enter") {
        filteredOptions[selectedItemIndex].onAction?.();
      }
    },
    [filteredOptions, selectedItemIndex]
  );

  useEffect(() => {
    addEventListener("keydown", onUp);
    addEventListener("keydown", onDown);
    addEventListener("keydown", onEnter);
    return () => {
      removeEventListener("keydown", onUp);
      removeEventListener("keydown", onDown);
      removeEventListener("keydown", onEnter);
    };
  }, [selectedItemIndex, filteredOptions]);

  return (
    <Palette
      options={filteredOptions}
      filterInputValue={filterInput}
      onFilterInputValueChange={setFilterInput}
      selectedItemIndex={selectedItemIndex}
      onSetSelectedItemIndex={setSelectedItemIndex}
    />
  );
};
