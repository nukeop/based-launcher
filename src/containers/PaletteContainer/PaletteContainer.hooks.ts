import { IpcEvent } from "../../../common/ipc";
import { ArgsContext } from "../../contexts/argsContext";
import { useDesktopEntries } from "../../hooks/useDesktopEntries";
import { useFlags } from "../../hooks/useFlags";
import { ipcRenderer } from "electron";
import Fuse from "fuse.js";
import { useCallback, useContext, useEffect, useRef, useState } from "react";
import { VariableSizeList } from "react-window";

export const usePaletteContainerProps = () => {
  const listRef = useRef<VariableSizeList>(null);
  const { stdinArgs, cliFlags, desktopEntries } = useContext(ArgsContext);

  const options =
    stdinArgs?.map((arg, index) => ({
      id: arg,
      name: arg,
      icon: null,
      onAction: () => onAction(arg),
    })) ?? [];

  // const options = desktopEntries?.map((entry) => ({
  //   id: entry["Desktop Entry"].Name,
  //   name: entry["Desktop Entry"].Name,
  //   description: entry["Desktop Entry"].Comment,
  //   icon: entry["Desktop Entry"].Icon,
  //   onAction: () => onAction(entry["Desktop Entry"].Exec),
  // }));

  const onAction = useCallback((item: string) => {
    ipcRenderer.send(IpcEvent.ReturnSelectedItem, item);
  }, []);

  const [filterInput, setFilterInput] = useState("");
  const [selectedItemIndex, setSelectedItemIndex] = useState(0);

  const fuse = new Fuse(options, {
    keys: ["name", "description"],
    threshold: 0.3,
  });

  const filteredOptions = filterInput
    ? fuse.search(filterInput).map((result) => result.item)
    : options;

  const onUp = useCallback(
    (event: KeyboardEvent) => {
      if (event.key === "ArrowUp") {
        setSelectedItemIndex(
          selectedItemIndex === 0
            ? filteredOptions.length - 1
            : selectedItemIndex - 1
        );
      }
    },
    [filteredOptions, selectedItemIndex]
  );

  const onDown = useCallback(
    (event: KeyboardEvent) => {
      if (event.key === "ArrowDown") {
        setSelectedItemIndex(
          selectedItemIndex === filteredOptions.length - 1
            ? 0
            : selectedItemIndex + 1
        );
      }
    },
    [filteredOptions, selectedItemIndex]
  );

  const onEnter = useCallback(
    (event: KeyboardEvent) => {
      if (event.key === "Enter") {
        filteredOptions[selectedItemIndex].onAction?.();
      }
    },
    [filteredOptions, selectedItemIndex]
  );

  useEffect(() => {
    listRef.current?.scrollToItem(selectedItemIndex);
  }, [selectedItemIndex]);

  useEffect(() => {
    addEventListener("keydown", onUp);
    addEventListener("keydown", onDown);
    addEventListener("keydown", onEnter);
    return () => {
      removeEventListener("keydown", onUp);
      removeEventListener("keydown", onDown);
      removeEventListener("keydown", onEnter);
    };
  }, [selectedItemIndex, filteredOptions]);

  return {
    listRef,
    flags: cliFlags,
    filteredOptions,
    filterInput,
    setFilterInput,
    selectedItemIndex,
    setSelectedItemIndex,
  };
};
