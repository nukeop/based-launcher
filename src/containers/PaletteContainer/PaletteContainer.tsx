import { Palette } from "../../components/Palette/Palette";
import { usePaletteContainerProps } from "./PaletteContainer.hooks";

export const PaletteContainer: React.FC = () => {
  const {
    listRef,
    flags,
    options,
    filterInput,
    setFilterInput,
    selectedItemIndex,
    setSelectedItemIndex,
  } = usePaletteContainerProps();

  return (
    <Palette
      options={options}
      filterInputValue={filterInput}
      onFilterInputValueChange={setFilterInput}
      selectedItemIndex={selectedItemIndex}
      onSetSelectedItemIndex={setSelectedItemIndex}
      listRef={listRef}
      prefixLabel={flags?.inputPrefix}
    />
  );
};
