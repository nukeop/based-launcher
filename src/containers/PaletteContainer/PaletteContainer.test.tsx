import { ArgsContext } from "../../contexts/argsContext";
import { AppRoot } from "../../layouts/AppRoot";
import { PaletteContainer } from "./PaletteContainer";
import {
  cleanup,
  fireEvent,
  render,
  RenderResult,
  waitFor,
} from "@testing-library/react";
import { CLIFlags } from "common/cliFlags";
import { LauncherOption } from "common/launcher";
import { anOption } from "common/tests/test-utils";
import { afterEach, describe, expect, it, vi } from "vitest";

vi.mock("react-virtualized-auto-sizer", () => ({
  default: ({ children }: any) => children({ height: 600, width: 600 }),
}));

vi.mock("electron", () => ({
  ipcRenderer: {
    invoke: () => Promise.resolve(),
  },
}));

const defaultOptions = [
  anOption({ id: "1", name: "first" }),
  anOption({ id: "2", name: "second" }),
  anOption({ id: "3", name: "third" }),
];

describe("Palette container", () => {
  afterEach(() => {
    cleanup();
  });

  it("renders a palette from piped arguments", async () => {
    const component = mountComponent(defaultOptions);

    waitFor(() => expect(component.asFragment()).toMatchSnapshot());
  });

  it("filters entries based on entered text", async () => {
    const component = mountComponent(defaultOptions);

    const input = (await component.findAllByTestId("filter-input")).at(0);
    if (input) {
      fireEvent.change(input, { target: { value: "fir" } });
    }

    waitFor(() => expect(component.getByText("first")).toBeTruthy());
    waitFor(() => expect(component.getByText("second")).toBeNull());

    component.unmount();
  });

  it("selects the first item by default", async () => {
    const component = mountComponent(defaultOptions);

    const firstItem = nthItem(component, "1");
    expect(firstItem).toHaveAttribute("data-selected", "true");
  });

  it("down arrow selects the next item", async () => {
    const component = mountComponent([
      ...defaultOptions,
      anOption({ id: "4", name: "fourth" }),
    ]);
    goDown(component);

    const secondItem = nthItem(component, "2");
    expect(secondItem).toHaveAttribute("data-selected", "true");
  });

  it("up arrow selects the previous item", async () => {
    const component = mountComponent([
      ...defaultOptions,
      anOption({ id: "4", name: "fourth" }),
    ]);

    goDown(component);
    const secondItem = nthItem(component, "2");
    expect(secondItem).toHaveAttribute("data-selected", "true");

    goUp(component);
    const firstItem = nthItem(component, "1");
    expect(firstItem).toHaveAttribute("data-selected", "true");
  });

  it("can use the wikipedia bang", async () => {
    fetchMock.mockOnce(
      JSON.stringify({
        query: {
          pages: {
            "123": {
              title: "Wikipedia",
              extract:
                "Wikipedia is a free online encyclopedia, created and edited by volunteers around the world and hosted by the Wikimedia Foundation.",
              pageprops: {
                "wikibase-shortdesc": "free online encyclopedia",
              },
            },
          },
        },
      })
    );
    const component = mountComponent(defaultOptions);

    const input = (await component.findAllByTestId("filter-input")).at(0);
    if (input) {
      fireEvent.change(input, { target: { value: "!w wikipedia" } });
    }

    expect(await component.findByText("Wikipedia")).toBeInTheDocument();
    expect(
      await component.findByText("free online encyclopedia")
    ).toBeInTheDocument();

    component.unmount();
  });

  it("shows that nothing was found when using the wikipedia bang", async () => {
    fetchMock.mockOnce(
      JSON.stringify({
        query: {
          pages: {
            "123": {},
          },
        },
      })
    );
    const component = mountComponent(defaultOptions);

    const input = (await component.findAllByTestId("filter-input")).at(0);
    if (input) {
      fireEvent.change(input, { target: { value: "!w wikipedia" } });
    }

    expect(await component.findByText("Nothing found")).toBeInTheDocument();

    component.unmount();
  });

  const mountComponent = (
    options: LauncherOption[] = [],
    flags: CLIFlags = {}
  ) => {
    return render(
      <ArgsContext.Provider value={{ flags, options, isLoading: true }}>
        <AppRoot>
          <PaletteContainer />
        </AppRoot>
      </ArgsContext.Provider>
    );
  };

  const goUp = (component: RenderResult) =>
    fireEvent.keyDown(component.getByTestId("palette"), { key: "ArrowUp" });

  const goDown = (component: RenderResult) =>
    fireEvent.keyDown(component.getByTestId("palette"), { key: "ArrowDown" });

  const nthItem = (component: RenderResult, n: string) =>
    component.getByTestId(`palette-item-${n}`);
});
