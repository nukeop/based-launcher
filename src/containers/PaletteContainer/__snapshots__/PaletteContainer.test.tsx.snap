// Vitest Snapshot v1

exports[`Palette container > renders a palette from piped arguments 1`] = `
<DocumentFragment>
  <div
    class="_app-root_z8vvx_1 app-root"
  >
    <div
      class="_palette_11q5b_1 palette"
      data-testid="palette"
      tabindex="-1"
    >
      <div
        class="_inputWrapper_rbxir_1 filter-input"
      >
        <input
          autocomplete="off"
          class="_input_rbxir_1"
          data-testid="filter-input"
          placeholder="Type to filter..."
          value=""
        />
      </div>
      <div
        class="_bangs-area_1rcuj_1 _empty_1rcuj_5"
      />
      <div
        class="_palette-grid_ch72j_1 _columns-1_ch72j_5 palette-grid"
      >
        <div
          class="_palette-grid-list_ch72j_17 palette-grid-list"
          style="position: relative; height: 600px; width: 600px; overflow: auto; will-change: transform; direction: ltr;"
        >
          <div
            style="height: 96px; width: 100%;"
          >
            <button
              class="_palette-item_102lw_1 _selected_102lw_17 palette-item--selected palette-item"
              data-selected="true"
              data-testid="palette-item-1"
              style="position: absolute; left: 0px; top: 0px; height: 32px; width: 100%;"
            >
              <div
                class="_palette-item-right_102lw_28"
              >
                <p>
                  first
                </p>
                <p
                  class="_palette-item-description_102lw_34"
                />
              </div>
            </button>
            <button
              class="_palette-item_102lw_1 palette-item"
              data-selected="false"
              data-testid="palette-item-2"
              style="position: absolute; left: 0px; top: 32px; height: 32px; width: 100%;"
            >
              <div
                class="_palette-item-right_102lw_28"
              >
                <p>
                  second
                </p>
                <p
                  class="_palette-item-description_102lw_34"
                />
              </div>
            </button>
            <button
              class="_palette-item_102lw_1 palette-item"
              data-selected="false"
              data-testid="palette-item-3"
              style="position: absolute; left: 0px; top: 64px; height: 32px; width: 100%;"
            >
              <div
                class="_palette-item-right_102lw_28"
              >
                <p>
                  third
                </p>
                <p
                  class="_palette-item-description_102lw_34"
                />
              </div>
            </button>
          </div>
        </div>
      </div>
      <footer
        class="_footer_uvi8w_1"
      >
        Type to filter, use bangs to search (e.g. 
        <kbd
          class="_key_1c8fc_1"
        >
          !w
        </kbd>
         to search Wikipedia).
      </footer>
    </div>
  </div>
</DocumentFragment>
`;
